[Segment Replication] Allocation and rebalancing based on average primary shard count per index

dreamer-89 opened the pull request:
### Description
This change adds average primary shard count per node constraint to both allocation and rebalancing operations and basically implements the approach 1 as [mentioned](https://github.com/opensearch-project/OpenSearch/issues/6210#issuecomment-1435032380) in #6210

### Changes
- Adds a new dynamic setting `cluster.routing.allocation.balance.prefer_primary` defaults to `false`. This setting defines if primary shard balance is needed or not.
- Adds a new constraint `isPrimaryShardsPerIndexPerNodeBreached` which breaches when a node contains more than average primary shards of an index. When breached, the node weight calculation adds `100k` to the node's weight resulting in lower chances for the node to be selected as target for unassigned shard allocation or relocation from rebalancing. Pleaset note, this is a soft limit which means even if node breaches constraint it can still be selected for allocation compared to hard limit from different allocation & rebalance deciders which removes node from allocation consideration.
- Restructure the constraint classes to consume changes to dynamic setting. When setting is enabled, constraints are applied during allocation and rebalancing
- Copies code from https://github.com/opensearch-project/OpenSearch/pull/6325 to move primaries first for balancing. This is essential to converge the balancing round quickly and minimize relocations.
- Unit and integration tests
- There are few tests which currently are flaky due to limitation called out in https://github.com/opensearch-project/OpenSearch/issues/6210#issuecomment-1435032380 where optimal primary shard distribution can't be attained due to `SameShardAllocationDecider` allocation decider. 


### Issues Resolved
#6210 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dreamer-89 commented:
@shwetathareja @nknize @Bukhtawar @vigyasharma : Need help with the review. 
CC @anasalkouz @mch2

mch2 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java:
```diff
@@ -8,58 +8,55 @@
-            this.index = index;
-        }
-    }
+    public final static long INDEX_SHARD_PER_NODE_BREACH_WEIGHT = 100000L;
```
I +1 @vigyasharma's comment here https://github.com/opensearch-project/OpenSearch/issues/6210#issuecomment-1444438462 around simplicity and comparing constraints.  On that note, can we keep the existing `CONSTRAINT_WEIGHT` value and move it to the Constraint class?

dreamer-89 replied:
Ack. Will use only a single weight constant for both allocation and rebalance constraints.

mch2 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java:
```diff
@@ -8,58 +8,55 @@
+        return params.weight(constraintSet);
+    }
+
+    public AllocationConstraints() {
```
Nit pls put constructor before other methods.

dreamer-89 replied:
Updated

mch2 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java:
```diff
@@ -8,58 +8,55 @@
+     * Weight calculation in other scenarios like shard movement and re-balancing remain unaffected by this constraint.
      */
+    public final static String INDEX_SHARD_PER_NODE_BREACH_CONSTRAINT_ID = "index.shard.breach.constraint";
+    private Map<String, Constraint> constraintSet;
```
nit, this is not a set: `private Map<String, Constraint> constraints;`

dreamer-89 replied:
Updated. 
Initially used the set that did not solve the purpose. Converted to map but use same naming variable.

mch2 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:
```diff
@@ -985,9 +1000,15 @@ private boolean tryRelocateShard(BalancedShardsAllocator.ModelNode minNode, Bala
                 if (allocationDecision.type() == Decision.Type.NO) {
                     continue;
                 }
+                // Skip moving the primary relocation when target has less primaries and is skewed due to
```
Not quite following this, could you pls explain?

dreamer-89 replied:
Thanks for calling this out. This change basically is to prevent un-necessary primary relocations from maxNode to minNode or in other words avoid blind relocations when moving primary from maxNode -> minNode is not desirable. Updated the comment.

mch2 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -264,18 +316,34 @@ static class WeightFunction {
             this.constraints = new AllocationConstraints();
+            this.rebalanceConstraints = new RebalanceConstraints();
+            // Enable index shard per node breach constraint
+            updateAllocationConstraint(INDEX_SHARD_PER_NODE_BREACH_CONSTRAINT_ID, true);
```
This is not dynamic, can we move enabling to AllocationConstraints?

dreamer-89 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/11750/
> * **CommitID:** [1033bda](https://github.com/opensearch-project/OpenSearch/commit/1033bdae5af6eca6e42ab4056a27b61671f2b3e4)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

The test fails avg primary shard count per index assertion because primary shard can't be moved due to`SameShardAllocationDecider`, exacerbated by auto expand replicas (`0-all`) sitting on all nodes. Will add this test once https://github.com/opensearch-project/OpenSearch/issues/6481 is implemented. 
```
routing_nodes:
-----node_id[r-WS5BibSHSl9MUITm0jYQ][V]
--------[test0][2], node[r-WS5BibSHSl9MUITm0jYQ], [P], s[STARTED], a[id=dOz4HlJoSVerANHdiWFX2w]
--------[test0][0], node[r-WS5BibSHSl9MUITm0jYQ], [P], s[STARTED], a[id=bfS_MHrHThCGkOsPBluW4w]
--------[test0][1], node[r-WS5BibSHSl9MUITm0jYQ], [R], s[STARTED], a[id=brTiZmNBQruIbJZLdfEcvw]
-----node_id[eO7OnBXWTqmwjITdka1uCQ][V]
--------[test0][1], node[eO7OnBXWTqmwjITdka1uCQ], [P], s[STARTED], a[id=vE1qbt4cTdWRGVN-eXD_4Q]
--------[test0][2], node[eO7OnBXWTqmwjITdka1uCQ], [R], s[STARTED], a[id=6fcIVtoOSJqu9WtaYGTZYg]
--------[test0][0], node[eO7OnBXWTqmwjITdka1uCQ], [R], s[STARTED], a[id=OiKu_9NcSH2Fy17QWSjYuw]
-----node_id[LKj2NBHJRfmWxbtJhiEy7w][V]
--------[test0][1], node[LKj2NBHJRfmWxbtJhiEy7w], [R], s[STARTED], a[id=NwHonQmURZi0cqwJ9pcFqg]
--------[test0][2], node[LKj2NBHJRfmWxbtJhiEy7w], [R], s[STARTED], a[id=t54DyrUJTTKV_AuWzb4HhQ]
--------[test0][0], node[LKj2NBHJRfmWxbtJhiEy7w], [R], s[STARTED], a[id=RdILEVjlS7SM_3wl-Zwz7Q]
```

mch2 commented on server/src/test/java/org/opensearch/cluster/routing/allocation/BalanceConfigurationTests.java:
```diff
@@ -494,5 +913,97 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
+/**
+ * Utility class to show shards distribution across nodes.
+ */
+class ShardAllocations {
```
This looks identical to the test component in SegmentReplicationAllocationIT - both tests are in the same project so we should be able to extract to a component.

dreamer-89 replied:
Yes you are right, though there is minor difference in how `RoutingNodes` are build in IT and Unit tests. I think it is possible to refactor this into a single class. Let me try to update this.

mch2 reviewed (commented):
@dreamer-89 made another pass through this & your tests.  Its looking good to me, few minor comments.

dreamer-89 commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationAllocationIT.java:
```diff
@@ -0,0 +1,364 @@
+            totalReplicaCount += replicaCount;
+            createIndex("test" + i, shardCount, replicaCount, i % 2 == 0);
+            logger.info("--> Creating index {} with shard count {} and replica count {}", "test" + i, shardCount, replicaCount);
+            assertBusy(() -> ensureGreen(), 60, TimeUnit.SECONDS);
```
Updated.

mch2 reviewed (approved):
LGTM, thanks @dreamer-89.

dreamer-89 commented:
> @shwetathareja @nknize @Bukhtawar @vigyasharma : Need help with the review. CC @anasalkouz @mch2

As this is core change, it will be good to have one more review. @shwetathareja @nknize @Bukhtawar @vigyasharma ping for review.

andrross commented on CHANGELOG.md:
```diff
@@ -72,6 +73,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Added
 - Add GeoTile and GeoHash Grid aggregations on GeoShapes. ([#5589](https://github.com/opensearch-project/OpenSearch/pull/5589))
 - Disallow multiple data paths for search nodes ([#6427](https://github.com/opensearch-project/OpenSearch/pull/6427))
+- [Segment Replication] Balanced primary shard distribution based allocation and rebalancing constraints ([#6422](https://github.com/opensearch-project/OpenSearch/pull/6422))
```
Why are there 2 changelog entries here? With different text? Assuming this will be backported to 2.x then a changelog entry should only go in the `Unreleased 2.x` section.

dreamer-89 replied:
Yes, my bad. Added one more changelog entry after removing the `skip-changelog` label. Adding a single entry in unrelease `2.x` section.

dreamer-89 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.search.SearchWeightedRoutingIT.testSearchAggregationWithNetworkDisruption_FailOpenEnabled
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/11881/
> * **CommitID:** [f1ef7fc](https://github.com/opensearch-project/OpenSearch/commit/f1ef7fc1c971441f4f05be84015873ab04b9dfd5)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Tracked in https://github.com/opensearch-project/OpenSearch/issues/5957

dreamer-89 commented:
Merging the changes now. 
@shwetathareja @nknize @Bukhtawar @vigyasharma : Please feel free to comment if you have any questions/concerns.

nknize commented:
No concerns, here. I haven't had a chance to dig deep on this but I did give it a quick once over. I like where this is going and great suggestion with the use of constraints. 

I think merging is fine so we can let it bake. I'll be giving another pass when I get back to teasing out the common server components.

dreamer-89 commented:
This pull request introduces significant changes to the OpenSearch project, focusing on segment replication, shard allocation, and balancing constraints. The updates include new tests for segment replication allocation, modifications to allocation constraints, and the introduction of constraints for rebalancing. Below are the most important changes:

### Segment Replication and Allocation Tests:
* Added a new test class `SegmentReplicationAllocationIT` to verify balanced primary shard allocation under various scenarios, including multiple indices creation, node disruptions, and single index shard allocation.

### Allocation Constraints:
* Refactored `AllocationConstraints` to use a `Map` for storing constraints and introduced methods to update allocation constraints dynamically. Added a new constraint for primary shard balance. [[1]](diffhunk://#diff-38fd557126b8726fed9e9914005f0854ac9ce89d5565f2a726f3ad31df600620L11-R52) [[2]](diffhunk://#diff-38fd557126b8726fed9e9914005f0854ac9ce89d5565f2a726f3ad31df600620L79-L86)
* Introduced a new `Constraint` class to encapsulate allocation constraint logic, including predicate evaluation and weight calculation.

### Rebalance Constraints:
* Added a new class `RebalanceConstraints` to handle constraints during rebalancing, including a constraint for primary shard balance per index per node.

### Balanced Shards Allocator:
* Updated `BalancedShardsAllocator` to integrate new allocation and rebalance constraints, ensuring primary shards are moved first during shard movements. [[1]](diffhunk://#diff-1473f4e5e9c712791e4971b58cda4dbfd99d09f912698f8f2d55cf49a01275bcR46) [[2]](diffhunk://#diff-1473f4e5e9c712791e4971b58cda4dbfd99d09f912698f8f2d55cf49a01275bcR62-R63) [[3]](diffhunk://#diff-1473f4e5e9c712791e4971b58cda4dbfd99d09f912698f8f2d55cf49a01275bcR99-R102)

### Documentation and Dependencies:
* Updated `CHANGELOG.md` to include the addition of segment replication allocation and rebalancing based on average primary shard count per index.

These changes enhance the robustness and flexibility of shard allocation and balancing mechanisms in OpenSearch, ensuring better distribution and management of primary shards across nodes.
